### DISCUSSION TITLE
File Scoped Namespace Class

### DISCUSSION THREAD

**Discussion by viniciusxdl** [dotnet/csharplang#7944] [Language Ideas]
C# is a wonderful language and I really love it so much, and something that I particularly like a lot is that it is a language that, in addition to being extremely powerful, does not stop in time and is always bringing new features, facilitating how we code and not getting stuck. behind new languages.

Something simply incredible were the top-level instructions, where instead of having to do something like:

``` csharp
using System;

namespace MyApp;

public class Program
{
    public static void Main(string[] args)
    {
        Console.WriteLine("Hello Word!");
    }
}
```

You can do just this:

``` csharp
Console.WriteLine("Hello Word!");
```

With that in mind, something I would like to suggest a new feature for a few reasons:

- Make it even easier and generate familiarity for new developers who came from other languages.
- Be useful for the class and file model that is generally widely used.
- Decrease a recoil.

It would be something like `File Scoped Namespace` but for classes. When used, only 1 class and 1 namespace could be defined in the entire file:

### Visual Studio 2022
![image](https://github.com/dotnet/csharplang/assets/22042217/1e898c9c-ecc6-4de7-b5de-fc41202f7a40)

### Code

``` csharp
namespace public class MyApp.Pages.Index;

public override Task OnInitializedAsync()
{
    // ...
}

public override Task OnParametersSetAsync()
{
    // ...
}
```

### Visual Studio
![image](https://github.com/dotnet/csharplang/assets/22042217/53d029e4-5965-4be1-97b7-cb6e4406bc0f)

### Code

``` csharp
namespace public class MyApp.Pages.Index;

[Paramater]
public string Code { get; init; }

public override Task OnInitializedAsync()
{
    // ...
}
```

### Visual Studio
![image](https://github.com/dotnet/csharplang/assets/22042217/2e73a202-8af7-48ef-9ed3-8d19c4f5b2da)

### Code
``` csharp
namespace public class MyApp.Code;

public Code(string code)
{
    // ...
}

public decimal GetDecodeResult()
{
    // ...
}
```

## Separate namespace and class

Another suggestion, if it is better to separate namespace and class:

### Visual Studio
![image](https://github.com/dotnet/csharplang/assets/22042217/be90fc10-242e-4b6a-a7ee-ead8ed114467)

### Code

``` csharp
public class Code with namespace MyApp.Models;

public Code(string code)
{
    // ...
}

public decimal GetDecodeResult()
{
    // ...
}
```

Or until:

### Visual Studio
![image](https://github.com/dotnet/csharplang/assets/22042217/142cbaba-026f-4def-bac8-5f5f5f619fcc)

### Code

``` csharp
namespace MyApp.Models with public class Code;

public Code(string code)
{
    // ...
}

public decimal GetDecodeResult()
{
    // ...
}
```

**Translated by Google Translate.**

> *EDIT:*

It ends up simplifying the language a lot, especially for beginners, by linking each class per file, making it more user-friendly, for example:

## No namespace

**Example.cs**
``` csharp
class Example;

public string GetExample()
{
}
```

**Program.cs**
``` csharp
var example = new Example();

Console.WriteLine(user.GetExample());
```

## With namespace

**Example.cs**
``` csharp
namespace class MyApp.Example;

public string GetExample()
{
}
```

**Program.cs**
``` csharp
using MyApp;

var example = new Example();
// Or:
var example = new MyApp.Example();

Console.WriteLine(user.GetExample());
```

The teaching is much better.

**Comment by sharpchen** [comment]
I don't see any benefit in this; it only eliminates two curly braces and one indentation level. An analyzer might be more suitable than a language feature for enforcing one type per file.

**Comment by viniciusxdl** [comment]
@sharpchen 

It ends up simplifying the language a lot, especially for beginners, by linking each class per file, making it more user-friendly, for example:

## No namespace

**Example.cs**
``` csharp
class Example;

public string GetExample()
{
}
```

**Program.cs**
``` csharp
var example = new Example();

Console.WriteLine(user.GetExample());
```

## With namespace

**Example.cs**
``` csharp
namespace class MyApp.Example;

public string GetExample()
{
}
```

**Program.cs**
``` csharp
using MyApp;

var example = new Example();
// Or:
var example = new MyApp.Example();

Console.WriteLine(user.GetExample());
```

The teaching is much better.

**Comment by 333fred** [comment]
Duplicate of https://github.com/dotnet/csharplang/discussions/928.
